README.md
#intimidate

intimidate is a node module to upload files to S3 with support for
automatic retry and exponential backoff.

It uses the excellent [knox](https://github.com/LearnBoost/knox) library to
handle the heavy lifting.

> When you need those uploads to back off, use *intimidate*™. - The Readme

## Installation

```bash
npm install intimidate
```

## Examples

Upload a local file:

```JavaScript
var Intimidate = require('intimidate')
// Create a client
var client = new Intimidate({
  key: 'some-aws-key',
  secret: 'some-aws-secret',
  bucket: 'lobsters',
  maxRetries: 5
})

client.upload('path/to/a/file.xml', 'destination/path/on/s3.xml', function(err, res) {
  if (err) {
    console.log('oh noes, all uploads failed! last error was', err)
  }
  else {
    console.log('yahoo, upload succeeded!')
  }
})
```

## API

### `Intimidate(opts)`

The constructor takes any opts that can be passed to
[knox's](https://github.com/LearnBoost/knox) `createClient` function. Here are
some important ones.

* `key` - S3 api key. Required.
* `secret` - S3 api secret. Required.
* `bucket` - S3 bucket to upload to. Required.
* `region` - S3 region to upload to. Defaults to `'us-west-2'`
* `maxRetries` - the number of times to retry before failing. Defaults to 3.
* `backoffInterval` a multiplier used to calculate exponential backoff. Larger
   numbers result in much larger backoff times after each failure. Defaults to 51.

Example:

```JavaScript
var Intimidate = require('intimidate')
var client = new Intimidate({
  key: 'love',
  secret: 'a sneaky secret',
  bucket: 'kicked',
  maxRetries: 4,
  backoffInterval: 20
})
```

### `upload(sourcePath, destination, cb)`


 Upload a file at sourcePath with automatic retries and exponential backoff

Params:

* `sourcePath` location of the file to upload on the fs
* `destination` path in s3 to upload file to
* `headers` (optional) object with any custom headers. `Content-Length` and `Content-Type` are determined by Intimidate automatically, and cannot be overwritten.
* `cb` function(err, res) called when upload is done or has
    failed too many times. `err` is the last error, and `res` is the reponse
    object if the request succeeded


Example:

```JavaScript
client.upload('a_car.zip', 'uploaded_cars/car.zip', { 'x-amz-acl': 'public-read' }, function(err, res) {
  if (err) {
    console.log('Dang, guess you can\'t upload a car.', err)
  }
  else {
    console.log('I uploaded a car.')
  }
})
```

### `uploadBuffer(buffer, headers, destination, cb)`

Upload a buffer

Params:

* `buffer` buffer to upload
* `headers` HTTP headers to set on request. `'Content-Length'` will default to
   `buffer.length`, and `'Content-Type'` will default to
   'application/octet-stream' if not provided.
* `destination` path on S3 to put file
* `cb` function(err, res) called when request completes or fails too many times


Example:

```JavaScript
var data = new Buffer('Shall I compare thee to a summer/'s day?')
var headers = {
  'Content-Type': 'application/text',
  'Content-Length': data.length
}

client.uploadBuffer(data, headers, 'poem_idea.txt', function(err, res) {
  if (err) {
    console.log('error uplaoding my sweet poem idea', err)
  }
  else {
    console.log('my poem idea is successfully archived to s3')
  }
})
```

### `uploadFiles(files, cb)`

Upload an array of files. The callback will be called when they all upload
successfully, or when at least one of the uploads has failed.

Params:

* `files` Array of `{src: 'some_path.file', dest: 'some_uploaded_path.file', headers: { 'x-amz-acl': 'public-read' } }`
  file object to be uploaded. Headers are optional.
* `cb` `function(err, res)` that will be called when upload is complete or
  one of the files has failed to upload.

Example:


```JavaScript
var files = [{src: 'hurp.txt', dest: 'durp.txt'}, {src: 'foo.txt', dest: 'foo.txt', headers: { 'x-amz-acl': 'public-read' } }]
client.uploadFiles(files, function(err, res) {
  if (err) {
    console.error('error uploading one file', err)
  }
  else {
    console.log('hooray, successfully uploaded all files')
  }
})
```


CHANGELOG.md
v1.1.0 - Thu, 26 Feb 2015 16:20:39 GMT
--------------------------------------

- [fa90305](../../commit/fa90305) [added] set custom headers for files


v1.0.0 - Tue, 23 Dec 2014 12:08:46 GMT
--------------------------------------

- 




test/index.js
/* eslint-env mocha */
var assert = require('assert')
var path = require('path')
var Retry = require('../index')

describe('Retry', function() {
  var noopKnox = {createClient: function() {}}

  it('throws an error if required options are not passed in', function() {
    assert.throws(function () {
      var client = new Retry({})
    })
  })

  it('does not throw an error if required options are passed in', function () {
    assert.doesNotThrow(function() {
      var client = new Retry({key: 1, secret: 1, bucket: 1}, noopKnox)
    })
  })

  describe('calculateBackoff', function() {
    it('returns a larger backoff for larger numbers of retries', function() {
      var client = new Retry({key: 1, secret: 1, bucket: 1}, noopKnox)
      var lessRetriesBackoff = client.calculateBackoff(1)
      var moreRetriesBackoff = client.calculateBackoff(10)
      assert(moreRetriesBackoff > lessRetriesBackoff)
    })
  })

  // THIS IS SPARTA
  var failKnox = {
    createClient: function() {
      return {
        put: function(put, destination) {
          return {
            eventHandlers: {},
            on: function(event, cb) {
              this.eventHandlers[event] = cb
            },
            end: function(data) {
              this.eventHandlers['error']('AN ERROR')
            }
          }
        }
      }
    }
  }

  var successKnox = {
    createClient: function() {
      return {
        put: function(put, destination) {
          return {
            eventHandlers: {},
            on: function(event, cb) {
              this.eventHandlers[event] = cb
            },
            end: function(data) {
              this.eventHandlers['response']({statusCode: 200})
            }
          }
        }
      }
    }
  }

  describe('upload', function() {
    it('tries uploading to s3 until maxRetries tries or it gets the hose again', function(done) {
      var client = new Retry({key: 1, secret: 1, bucket: 1, backoffInterval: 1, maxRetries:4 }, failKnox)
      client.upload(path.join(__dirname, 'fakeFile.txt'), 'destination', function(err, res, timesRetried) {
        assert(err)
        assert(res == null)
        assert.equal(timesRetried, 4)
        done()
      })
    })

    it('tries uploading to s3 until maxRetries tries or it gets the hose again: passing headers', function(done) {
      var client = new Retry({key: 1, secret: 1, bucket: 1, backoffInterval: 1, maxRetries: 4 }, failKnox)
      var headers = { 'Content-Type': 'application/text' }
      client.upload(path.join(__dirname, 'fakeFile.txt'), 'destination', headers, function(err, res, timesRetried) {
        assert(err)
        assert(res == null)
        assert.equal(timesRetried, 4)
        done()
      })
    })

    it('calls the callback with a response object if the request succeeds', function(done) {
      var client = new Retry({key: 1, secret: 1, bucket: 1 }, successKnox)
      client.upload(path.join(__dirname, 'fakeFile.txt'), 'destination', function(err, res) {
        assert.ifError(err)
        assert(res)
        done()
      })
    })
  })

  it('calls the callback with a response object if the request succeeds: passing headers', function(done) {
    var client = new Retry({key: 1, secret: 1, bucket: 1 }, successKnox)
    var headers = { 'Content-Type': 'application/text' }
    client.upload(path.join(__dirname, 'fakeFile.txt'), 'destination', headers, function(err, res) {
      assert.ifError(err)
      assert(res)
      done()
    })
  })

  describe('uploadBuffer', function() {
    it('uploads a buffer', function(done) {
      var data = new Buffer('Shall i compare thee to a summer\'s day?')
      var headers = {
        'Content-Type': 'application/text'
      }

      var client = new Retry({key: 1, secret: 1, bucket: 1}, successKnox)
      client.uploadBuffer(data, headers, 'poem.txt', function(err, res) {
        assert.ifError(err)
        assert(res)
        done()
      })
    })
  })

  describe('uploadFiles', function() {
    it('calls the callback with a array response object if the request succeeds', function(done) {
      var client = new Retry({key: 1, secret: 1, bucket: 1 }, successKnox)
      var files = [{
        src: path.join(__dirname, 'fakeFile.txt'),
        dest: 'destination'
      },{
        src: path.join(__dirname, 'fakeFile.txt'),
        dest: 'another_destination'
      }];

      client.uploadFiles(files, function(err, res) {
        assert.ifError(err)
        assert(res)
        assert(res.length == files.length)
        done()
      })
    })
    it('calls the callback with the first error ', function(done) {
      var client = new Retry({key: 1, secret: 1, bucket: 1 }, successKnox)
      var files = [{
        src: path.join(__dirname, 'doesNotExist.txt'),
        dest: 'destination'
      },{
        src: path.join(__dirname, 'fakeFile.txt'),
        dest: 'another_destination'
      }];

      client.uploadFiles(files, function(err, res) {
        assert(err)
        assert(res)
        done()
      })
    })
  })
})


index.js
var fs = require('fs')
var knox = require('knox')
var mime = require('mime')
var waitress = require('waitress')

/**
 * @class Retry
 * @constructor
 * @param opts {Object} an object containing config parameters
 *   `key`, `secret` and `bucket` are all requried.
 *   `region` is the S3 region. It defaults to us-west-2
 *   `maxRetries` is the number of times to retry before failing. It defaults to 3
 *   `backoffInterval` is a multiplier used to calculate exponential backoff.
 *   It defaults to 51.
 * @param knox {Object} knox library by default, or mocked version for testing
 * @param mime {Object} mime lib by default, or mocked version for testing
 */
function Retry(opts, s3Lib, mimeLib) {
  // Inject deps for testing
  if (typeof s3Lib === 'undefined') {
    s3Lib = knox
  }
  this.knox = s3Lib

  if (typeof mimeLib === 'undefined') {
    mimeLib = mime
  }
  this.mime = mimeLib


  validateRequiredOptions(opts)

  this.key = opts.key
  this.secret = opts.secret
  this.bucket = opts.bucket
  this.region = opts.region || 'us-west-2'

  this.s3Client = this.knox.createClient(opts)

  this.maxRetries = opts.maxRetries || 3
  this.backoffInterval = opts.backoffInterval || 51
}

function validateRequiredOptions(opts) {
  if (!opts.hasOwnProperty('key')) {
    throw new Error("Missing required 'key' option from opts.")
  }
  else if (!opts.hasOwnProperty('secret')) {
    throw new Error("Missing required 'secret' option from opts.")
  }
  else if (!opts.hasOwnProperty('bucket')) {
    throw new Error("Missing required 'bucket' option from opts.")
  }
}

Retry.prototype.calculateBackoff = function(numRetries) {
  var randMultiplier = Math.ceil(Math.random() * (Math.pow(2, numRetries + 2) - 1))
  return this.backoffInterval * randMultiplier
}

/**
 * Upload a file at sourcePath with automatic retries and exponential backoff
 *
 * @param sourcePath {String} location of the file to upload on the fs
 * @param destination {String} path in s3 to upload file to
 * @param _headers {Object} (optional) custom headers to pass around.
 * @param cb {Function} function(err) called when upload is done or has failed too many times
 */
Retry.prototype.upload = function(sourcePath, destination, _headers, cb) {
  var self = this

  if (typeof _headers === 'function' && typeof cb === 'undefined') {
    cb = _headers;
    _headers = {};
  }

  fs.readFile(sourcePath, function(err, file) {
    if (err) {
      return cb(err)
    }

    var headers = {
      'Content-Type': self.mime.lookup(sourcePath),
      'Content-Length': file.length
    }

    // Add any properties from _headers to headers.
    // Don't override Content-Type and Content-Length
    for (var prop in _headers) {
      if (_headers.hasOwnProperty(prop) && prop.toLowerCase() != 'content-type' && prop.toLowerCase() != 'content-length') {
        headers[prop] = _headers[prop];
      }
    }

    self.uploadWithRetries(file, headers, destination, cb)
  })
}

/**
 * Upload a buffer with accompanying headers to S3.
 *
 * @param buffer {Buffer} buffer to put to s3
 * @param headers {Object} headers. Will set default Content-Type and
 *   Content-Length if none is provided.
 * @param destination {String} path to put buffer to
 * @param cb {Funtion} function(err, res) called when upload has succeeded
 *   or failed too many times.
 */
Retry.prototype.uploadBuffer = function(buffer, headers, destination, cb) {
  this.uploadWithRetries(buffer, headers, destination, cb)
}

/**
 * Upload a buffer with accompanying headers to s3. Recursively calls itself
 * until `timesRetried` exceeds `this.maxRetries`.
 *
 * @private
 * @param data {Buffer} data to put to S3
 * @param headers {Object} headers to send with request to S3. Will set a default
 *   Content-Length and Content-Type if none is provided.
 * @param destination {String} path to put the buffer to S3
 * @param timesRetried {Number} number of times this current upload has retried.
 *   Defaults to 0 if not passed in, and will increment every time an upload fails.
 * @param cb {Function} function(err, res) called when upload is done or has failed
 *   too many times.
 */
Retry.prototype.uploadWithRetries = function(data, headers, destination, timesRetried, cb) {
  var self = this
  // prevent callback from being called twice
  var callbackCalled = false

  // sometimes knox failures will give EPIPE errors after sending a non-200
  // status code, so we have to guard against recusing twice
  var recursionScheduled = false

  // Set content type and length if they aren't included
  headers = headers || {}
  if (!(headers['Content-Type'] || headers['content-type'])) {
    headers['Content-Type'] = 'application/octet-stream'
  }
  if (headers['Content-Length'] == undefined || headers['content-length'] == undefined) {
    headers['Content-Length'] = data.length
  }

  if (typeof timesRetried === 'function') {
    cb = timesRetried
    // start at -1 so we can increment up to 0 on the first call
    timesRetried = -1
  }

  timesRetried++

  function endWithError(err) {
    if (callbackCalled) {
      return
    }
    if (timesRetried >= self.maxRetries) {
      if (!callbackCalled) {
        callbackCalled = true
        return cb(err, null, timesRetried)
      }
    }
    else {
      if (recursionScheduled) {
        return
      }
      recursionScheduled = true
      setTimeout(self.uploadWithRetries.bind(self),
        self.calculateBackoff(timesRetried), data, headers, destination,
        timesRetried, cb)
    }
  }

  function endWithResponse(res) {
    if (res.statusCode !== 200) {
      return endWithError(new Error('Invalid status code: ' + res.statusCode))
    }

    if (!callbackCalled) {
      callbackCalled = true
      return cb(null, res, timesRetried)
    }
  }

  var client = this.s3Client.put(destination, headers)
  client.on('response', endWithResponse)
  client.on('error', endWithError)

  client.end(data)
}

/**
 * Upload a file at sourcePath with automatic retries and exponential backoff
 *
 * @param files {Object} { src: /path, dest: /path, headers:{} } location, destination and headers of the file to upload on the fs
 * @param cb {Function} function(err) called when all uploads are done or have failed too many times
 */
Retry.prototype.uploadFiles = function(files, cb) {
  var self = this
  var done = waitress(files.length, cb);

  files.forEach(function(file) {
    if(typeof file.headers === 'object' && file.headers !== null) {
      self.upload(file.src, file.dest, file.headers, done);
    } else {
      self.upload(file.src, file.dest, done);
    }
  });
}

module.exports = Retry
